lib/components/metrics.js

```diff
@@ -0,0 +1,293 @@
+ *       return new Thing();
+ *   }
+ *
+ * @constructor
+ */
+function Metrics() {
+    // Only attempt to load these dependencies if metrics are enabled
+    var Prometheus = require("prometheus-client");
+
+    var client = this._client = new Prometheus();
+
+    this._collectors = []; // executed in order
+    this._counters = {};
+
+    // Register some built-in process-wide metrics
+    // See also
+    //   https://prometheus.io/docs/instrumenting/writing_clientlibs/#standard-and-runtime-collectors
+
+    var rss_gauge = this.addGauge({
+        namespace: "process",
+        name: "resident_memory_bytes",
+        help: "Resident memory size in bytes",
+    });
+    var vsz_gauge = this.addGauge({
+        namespace: "process",
+        name: "virtual_memory_bytes",
+        help: "Virtual memory size in bytes",
+    });
+
+    var heap_size_gauge = this.addGauge({
+        namespace: "process",
+        name: "heap_bytes",
+        help: "Total size of Node.js heap in bytes",
+    });
+    var heap_used_gauge = this.addGauge({
+        namespace: "nodejs",
+        name: "heap_used_bytes",
+        help: "Used size of Node.js heap in bytes",
+    });
+
+    var cpu_gauge = this.addGauge({
+        namespace: "process",
+        name: "cpu_seconds_total",
+        help: "Total user and system CPU time spent in seconds",
+    });
+
+    var cpu_user_gauge = this.addGauge({
+        namespace: "process",
+        name: "cpu_user_seconds_total",
+        help: "Total user CPU time spent in seconds",
+    });
+    var cpu_system_gauge = this.addGauge({
+        namespace: "process",
+        name: "cpu_system_seconds_total",
+        help: "Total system CPU time spent in seconds",
+    });
+
+    this.addCollector(function() {
+        var usage = process.memoryUsage();
+
+        rss_gauge.set({}, usage.rss);
+        heap_size_gauge.set({}, usage.heapTotal);
+        heap_used_gauge.set({}, usage.heapUsed);
+
+        var stats = _read_proc_self_stat();
+
+        // CPU times in ticks
+        var utime_secs = stats[11] / TICKS_PER_SEC;
+        var stime_secs = stats[12] / TICKS_PER_SEC;
+
+        cpu_gauge.set({}, utime_secs + stime_secs);
+        cpu_user_gauge.set({}, utime_secs);
+        cpu_system_gauge.set({}, stime_secs);
+
+        // Virtual memory size
+        vsz_gauge.set({}, stats[20]);
+    });
+
+    this.addGauge({
+        namespace: "process",
+        name: "open_fds",
+        help: "Number of open file descriptors",
+        refresh: function(gauge) {
+            var fds = fs.readdirSync("/proc/self/fd");
+
+            // subtract 1 due to readdir handle itself
+            gauge.set(null, fds.length - 1);
+        }
+    });
+
+    this.addGauge({
+        namespace: "process",
+        name: "max_fds",
+        help: "Maximum number of open file descriptors allowed",
+        refresh: function(gauge) {
+            var limits = fs.readFileSync("/proc/self/limits");
```

Comment:
Why are we actively blocking the event loop whilst reading this? Why aren't we using the async form?